bigquery/src/main/java/org/apache/iceberg/gcp/bigquery/BigQueryTableOperations.java

```diff
@@ -0,0 +1,296 @@
+    String metadataLocation = null;
+    try {
+      metadataLocation =
+          loadMetadataLocationOrThrow(
+              client.load(this.tableReference).getExternalCatalogTableOptions());
+    } catch (NoSuchTableException e) {
+      if (currentMetadataLocation() != null) {
+        // Re-throws the exception because the table must exist in this case.
+        throw e;
+      }
+    }
+    refreshFromMetadataLocation(metadataLocation);
+  }
+
+  // The doCommit method should provide implementation on how to update with metadata location
+  // atomically
+  @Override
+  public void doCommit(TableMetadata base, TableMetadata metadata) {
+    String newMetadataLocation =
+        base == null && metadata.metadataFileLocation() != null
+            ? metadata.metadataFileLocation()
+            : writeNewMetadata(metadata, currentVersion() + 1);
+    BaseMetastoreOperations.CommitStatus commitStatus =
+        BaseMetastoreOperations.CommitStatus.FAILURE;
+    try {
+      if (base == null) {
+        createTable(newMetadataLocation, metadata);
+      } else {
+        updateTable(base.metadataFileLocation(), newMetadataLocation, metadata);
+      }
+      commitStatus = BaseMetastoreOperations.CommitStatus.SUCCESS;
+    } catch (CommitFailedException | CommitStateUnknownException e) {
+      throw e;
+    } catch (Throwable e) {
+      LOG.error("Exception thrown on commit: ", e);
+      if (e instanceof AlreadyExistsException) {
+        throw e;
+      }
+      commitStatus =
+          BaseMetastoreOperations.CommitStatus.valueOf(
+              checkCommitStatus(newMetadataLocation, metadata).name());
+      if (commitStatus == BaseMetastoreOperations.CommitStatus.FAILURE) {
+        throw new CommitFailedException(e, "Failed to commit");
+      }
+      if (commitStatus == BaseMetastoreOperations.CommitStatus.UNKNOWN) {
+        throw new CommitStateUnknownException(e);
+      }
+    } finally {
+      try {
+        if (commitStatus == BaseMetastoreOperations.CommitStatus.FAILURE) {
+          LOG.warn("Failed to commit updates to table {}", tableName());
+          io().deleteFile(newMetadataLocation);
+        }
+      } catch (RuntimeException e) {
+        LOG.error(
+            "Failed to cleanup metadata file at {} for table {}",
+            newMetadataLocation,
+            tableName(),
+            e);
+      }
+    }
+  }
+
+  @Override
+  public String tableName() {
+    return String.format("%s.%s", tableReference.getDatasetId(), tableReference.getTableId());
+  }
+
+  @Override
+  public FileIO io() {
+    return fileIO;
+  }
+
+  private void createTable(String newMetadataLocation, TableMetadata metadata) {
+    LOG.debug("Creating a new Iceberg table: {}", tableName());
+    Table tableBuilder = makeNewTable(metadata, newMetadataLocation);
+    tableBuilder.setTableReference(this.tableReference);
+    addConnectionIfProvided(tableBuilder, metadata.properties());
+
+    client.create(tableBuilder);
```

Comment:
Do we have any checks to ensure the namespace exists or that the proper exception will be thrown?